parquet-protobuf/src/main/java/org/apache/parquet/proto/ProtoWriteSupport.java

```diff
@@ -130,13 +137,36 @@ public WriteContext init(Configuration configuration) {
 
     this.messageWriter = new MessageWriter(messageDescriptor, rootSchema);
 
-    Map<String, String> extraMetaData = new HashMap<String, String>();
+    Map<String, String> extraMetaData = new HashMap<>();
     extraMetaData.put(ProtoReadSupport.PB_CLASS, protoMessage.getName());
     extraMetaData.put(ProtoReadSupport.PB_DESCRIPTOR, serializeDescriptor(protoMessage));
     extraMetaData.put(PB_SPECS_COMPLIANT_WRITE, String.valueOf(writeSpecsCompliant));
     return new WriteContext(rootSchema, extraMetaData);
   }
 
+  @Override
+  public FinalizedWriteContext finalizeWrite() {
+    Map<String, String> protoMetadata = new HashMap<>();
+    for (Map.Entry<String, Map<String, Integer>> enumNameNumberMapping : protoEnumBookKeeper.entrySet()) {
+      StringBuilder nameNumberPairs = new StringBuilder();
+      if (enumNameNumberMapping.getValue().isEmpty()) {
+        // No enum is ever written to any column of this file, put an empty string as the value in the metadata
+        LOG.info("No enum is written for " + enumNameNumberMapping.getKey());
+      }
+      int idx = 0;
+      for (Map.Entry<String, Integer> nameNumberPair : enumNameNumberMapping.getValue().entrySet()) {
```

Comment:
nit: since Parquet already depends on Guava, we can use Guava's [Joiner](https://google.github.io/guava/releases/27.0.1-jre/api/docs/com/google/common/base/Joiner.html) class to stringify the map to spare this for loop, something like this:
```
Joiner.on(METADATA_ENUM_ITEM_SEPARATOR)
        .withKeyValueSeparator(METADATA_ENUM_KEY_VALUE_SEPARATOR)
        .join(enumNameNumberMapping.getValue())
```
I couldn't find anything to simplify the reading path. Just an idea, maybe it is better not to depend on 3rd party libs too much. :)

Reply:
I would prefer to not use guava if possible

Reply:
In order to make the method easier to read, if we're not going with guava (which I do like, though I don't see it being used in `parquet-proto`), we could extract the joining part into a dedicated method. Something like:

```java
  @Override
  public FinalizedWriteContext finalizeWrite() {
    Map<String, String> protoMetadata = new HashMap<>();
    for (Map.Entry<String, Map<String, Integer>> enumNameNumberMapping : protoEnumBookKeeper.entrySet()) {
      
      if (enumNameNumberMapping.getValue().isEmpty()) {
        // No enum is ever written to any column of this file, put an empty string as the value in the metadata
        LOG.info("No enum is written for " + enumNameNumberMapping.getKey());
      }
      
      String nameNumberPairs = joinNameNumberPairs(enumNameNumberMapping.getValue());
      protoMetadata.put(METADATA_ENUM_PREFIX + enumNameNumberMapping.getKey(), nameNumberPairs);
    }
    return new FinalizedWriteContext(protoMetadata);
  }

  public String getNameNumberPairs(Map<String, Integer> nameNumberMap) {
    StringBuilder nameNumberPairs = new StringBuilder();
    int idx = 0;
    for (Map.Entry<String, Integer> nameNumberPair : nameNumberMap.entrySet()) {
      nameNumberPairs.append(nameNumberPair.getKey())
        .append(METADATA_ENUM_KEY_VALUE_SEPARATOR)
        .append(nameNumberPair.getValue());
      idx ++;
      if (idx < nameNumberMap.size()) {
        nameNumberPairs.append(METADATA_ENUM_ITEM_SEPARATOR);
      }
    }
    return nameNumberPairs;
  }
```

What do you think?

Reply:
Sorry for the late reply. @costimuraru
I refactor out all the logic of building enum metadata in a method. I think this would be more clear.